components/brave_wallet/browser/solana_tx_manager.cc

```diff
@@ -747,54 +927,119 @@ void SolanaTxManager::OnGetAccountInfo(
+      chain_id, base64_encoded_message,
       base::BindOnce(&SolanaTxManager::OnGetFeeForMessage,
-                     weak_ptr_factory_.GetWeakPtr(), std::move(callback)));
+                     weak_ptr_factory_.GetWeakPtr(), std::move(callback),
+                     std::move(meta), unsigned_tx));
 }
 
-void SolanaTxManager::OnGetFeeForMessage(GetEstimatedTxFeeCallback callback,
+void SolanaTxManager::OnGetFeeForMessage(GetEstimatedTxBaseFeeCallback callback,
+                                         std::unique_ptr<SolanaTxMeta> meta,
+                                         const std::string& unsigned_tx,
                                          uint64_t tx_fee,
                                          mojom::SolanaProviderError error,
                                          const std::string& error_message) {
-  std::move(callback).Run(tx_fee, error, error_message);
+  std::move(callback).Run(std::move(meta), unsigned_tx, tx_fee, error,
+                          error_message);
+}
+
+void SolanaTxManager::GetSolanaTxFeeEstimation(
+    const std::string& chain_id,
+    const std::string& tx_meta_id,
+    GetSolanaTxFeeEstimationCallback callback) {
+  // Get the TxMeta.
+  std::unique_ptr<SolanaTxMeta> meta =
+      GetSolanaTxStateManager()->GetSolanaTx(tx_meta_id);
+  if (!meta) {
+    std::move(callback).Run(
+        {}, mojom::SolanaProviderError::kInternalError,
+        l10n_util::GetStringUTF8(IDS_BRAVE_WALLET_TRANSACTION_NOT_FOUND));
+    return;
+  }
+
+  auto internal_callback =
+      base::BindOnce(&SolanaTxManager::FinishGetSolanaTxFeeEstimation,
+                     weak_ptr_factory_.GetWeakPtr(), std::move(callback));
+
+  GetSolanaTxFeeEstimationForMeta(chain_id, std::move(meta),
+                                  std::move(internal_callback));
+}
+
+void SolanaTxManager::GetSolanaTxFeeEstimationForMeta(
+    const std::string& chain_id,
+    std::unique_ptr<SolanaTxMeta> meta,
+    GetSolanaTxFeeEstimationForMetaCallback callback) {
+  // This function makes three RPC calls to create the SolanaFeeEstimation:
+  //
+  // 1. getFeeForMessage to get the base fee
+  // 2. simulateTransaction to get the compute unit estimation
+  // 3. getRecentPrioritizationFees to get the fee per compute unit
+  //
+  // If fetching the base fee fails, we set the base fee, compute units, and fee
+  // per compute unit to zero so the caller can identify the failure.
+  //
+  // If estimating the compute units fails, we set the correct base fee, but
+  // set the compute units and fee per compute unit to zero.
+  //
+  // If fetching the fee per compute unit fails, we set the correct base fee and
+  // compute units, but set the fee per compute unit to the default value.
+  auto internal_callback = base::BindOnce(
+      &SolanaTxManager::OnGetEstimatedTxBaseFee, weak_ptr_factory_.GetWeakPtr(),
+      chain_id, std::move(callback));
+  GetEstimatedTxBaseFee(std::move(meta), std::move(internal_callback));
```

Comment:
As mentioned in the DM, it seems we don't really have to chain all these as sequential operations and callbacks. We can run these three concurrently after we have blockhash info (either tx already has it or finish fetch it from network) and unsigned tx. This may have performance improvement and better readability.
I'm okay with it being a follow-up.